### PR TITLE
Referencing request from response

### LINKED ISSUE #237: Add an option to reference request from response
For a DSL looking more or less like this

```
Contract contractDsl = Contract.make {
				request {
					method 'GET'
					url('/api/v1/xxxx') {
						queryParameters {
							parameter("foo", "bar")
							parameter("foo", "bar2")
						}
					}
					headers {
						header(authorization(), "secret")
					}
					body(foo: "bar", baz: 5)
				}
				response {
					status 200
					headers {
						header(authorization(), fromRequest().header(authorization()))
					}
					body(
							url: fromRequest().url(),
							param: fromRequest().query("foo"),
							paramIndex: fromRequest().query("foo", 1),
							authorization: fromRequest().header("Authorization"),
							fullBody: fromRequest().body(),
							responseFoo: fromRequest().body('$.foo'),
							responseBaz: fromRequest().body('$.baz')
					)
				}
			}
```

we'd like to have a test that looks sth like this

```
 given:
  def request = given()
    .header("Authorization", "secret")
    .body('''{"foo":"bar"}''')

 when:
  def response = given().spec(request)
    .queryParam("foo","bar")
    .queryParam("foo","bar2")
    .get("/api/v1/xxxx")

 then:
  response.statusCode == 200
  response.header('Authorization')  == 'secret'
 and:
  DocumentContext parsedJson = JsonPath.parse(response.body.asString())
  assertThatJson(parsedJson).field("url").isEqualTo("/api/v1/xxxx")
  assertThatJson(parsedJson).field("fullBody").isEqualTo("{\"foo\":\"bar\",\"baz\":5}")
  assertThatJson(parsedJson).field("paramIndex").isEqualTo("bar2")
  assertThatJson(parsedJson).field("responseFoo").isEqualTo("bar")
  assertThatJson(parsedJson).field("authorization").isEqualTo("secret")
  assertThatJson(parsedJson).field("responseBaz").isEqualTo(5)
  assertThatJson(parsedJson).field("param").isEqualTo("bar")
```

and a stub that can pick part of the request and properly reference it in the response.

Any feedback on the API is more than welcome

### PR DESCRIPTION
**Pull request by marcingrzejszczak**
The best situation is to provide fixed values but sometimes you need to reference a request in your response. In order to do this you can profit from the `fromRequest()` method that allows you to reference a bunch of elements from the HTTP request. You can use the following options:

- `fromRequest().url()` - return the request URL
- `fromRequest().query(String key)` - return the first query parameter with a given name
- `fromRequest().query(String key, int index)` - return the nth query parameter with a given name
- `fromRequest().header(String key)` - return the first header with a given name
- `fromRequest().header(String key, int index)` - return the nth header with a given name
- `fromRequest().body()` - return the full request body
- `fromRequest().body(String jsonPath)` - return the element from the request that matches the JSON Path

Let’s take a look at the following contract

```groovy
Contract contractDsl = Contract.make {
	request {
		method 'GET'
		url('/api/v1/xxxx') {
			queryParameters {
				parameter("foo", "bar")
				parameter("foo", "bar2")
			}
		}
		headers {
			header(authorization(), "secret")
			header(authorization(), "secret2")
		}
		body(foo: "bar", baz: 5)
	}
	response {
		status 200
		headers {
			header(authorization(), "foo ${fromRequest().header(authorization())} bar")
		}
		body(
				url: fromRequest().url(),
				param: fromRequest().query("foo"),
				paramIndex: fromRequest().query("foo", 1),
				authorization: fromRequest().header("Authorization"),
				authorization2: fromRequest().header("Authorization", 1),
				fullBody: fromRequest().body(),
				responseFoo: fromRequest().body('$.foo'),
				responseBaz: fromRequest().body('$.baz'),
				responseBaz2: "Bla bla ${fromRequest().body('$.foo')} bla bla"
		)
	}
}
```

Running a JUnit test generation will lead in creation of a test looking more or less like this

```java
// given:
 MockMvcRequestSpecification request = given()
   .header("Authorization", "secret")
   .header("Authorization", "secret2")
   .body("{\"foo\":\"bar\",\"baz\":5}");

// when:
 ResponseOptions response = given().spec(request)
   .queryParam("foo","bar")
   .queryParam("foo","bar2")
   .get("/api/v1/xxxx");

// then:
 assertThat(response.statusCode()).isEqualTo(200);
 assertThat(response.header("Authorization")).isEqualTo("foo secret bar");
// and:
 DocumentContext parsedJson = JsonPath.parse(response.getBody().asString());
 assertThatJson(parsedJson).field("url").isEqualTo("/api/v1/xxxx");
 assertThatJson(parsedJson).field("fullBody").isEqualTo("{\"foo\":\"bar\",\"baz\":5}");
 assertThatJson(parsedJson).field("paramIndex").isEqualTo("bar2");
 assertThatJson(parsedJson).field("responseFoo").isEqualTo("bar");
 assertThatJson(parsedJson).field("authorization2").isEqualTo("secret2");
 assertThatJson(parsedJson).field("responseBaz").isEqualTo(5);
 assertThatJson(parsedJson).field("responseBaz2").isEqualTo("Bla bla bar bla bla");
 assertThatJson(parsedJson).field("param").isEqualTo("bar");
 assertThatJson(parsedJson).field("authorization").isEqualTo("secret");
```

As you can see elements from the request have been properly referenced in the response.

The generated WireMock stub will look more or less like this:

```json
{
  "request" : {
    "urlPath" : "/api/v1/xxxx",
    "method" : "POST",
    "headers" : {
      "Authorization" : {
        "equalTo" : "secret2"
      }
    },
    "queryParameters" : {
      "foo" : {
        "equalTo" : "bar2"
      }
    },
    "bodyPatterns" : [ {
      "matchesJsonPath" : "$[?(@.baz == 5)]"
    }, {
      "matchesJsonPath" : "$[?(@.foo == 'bar')]"
    } ]
  },
  "response" : {
    "status" : 200,
    "body" : "{\"url\":\"{{{request.url}}}\",\"param\":\"{{{request.query.foo.[0]}}}\",\"paramIndex\":\"{{{request.query.foo.[1]}}}\",\"authorization\":\"{{{request.headers.Authorization.[0]}}}\",\"authorization2\":\"{{{request.headers.Authorization.[1]}}}\",\"fullBody\":\"{{{escapejsonbody}}}\",\"responseFoo\":\"{{{jsonpath this '$.foo'}}}\",\"responseBaz\":{{{jsonpath this '$.baz'}}} ,\"responseBaz2\":\"Bla bla {{{jsonpath this '$.foo'}}} bla bla\"}",
    "headers" : {
      "Authorization" : "{{{request.headers.Authorization.[0]}}}"
    },
    "transformers" : [ "response-template" ]
  }
}
```

So sending a request as the one presented in the request part of the contract will lead in sending the following response body

```json
{
  "url" : "/api/v1/xxxx?foo=bar&foo=bar2",
  "param" : "bar",
  "paramIndex" : "bar2",
  "authorization" : "secret",
  "authorization2" : "secret2",
  "fullBody" : "{\"foo\":\"bar\",\"baz\":5}",
  "responseFoo" : "bar",
  "responseBaz" : 5,
  "responseBaz2" : "Bla bla bar bla bla"
}
```

IMPORTANT:

This feature will work only with WireMock having version greater or equal to 2.5.1. We’re using WireMock’s `response-template` response transformer. It’s using Handlebars to convert the Mustache `{{{ }}}` templates into proper values. Additionally we’re registering 2 helper functions. `escapejsonbody` - that escapes the request body in a format that can be embedded in a JSON. Another is `jsonpath` that for a given parameter knows how to find an object in the request body.

fixes #237

### CODE REVIEW

**Comment by cah-andrewfitzgerald** [comment]
Looks good to me.
The only thing i can think of is that it would be nice if the response body had nicer formatting in the Wiremock stub, but that would probably be more realistic if it wasn't an escaped string.


```
  "body" : "{\"url\":\"{{{request.url}}}\",\"param\":\"{{{request.query.foo.[0]}}}\",\"paramIndex\":\"{{{request.query.foo.[1]}}}\",\"authorization\":\"{{{request.headers.Authorization.[0]}}}\",\"authorization2\":\"{{{request.headers.Authorization.[1]}}}\",\"fullBody\":\"{{{escapejsonbody}}}\",\"responseFoo\":\"{{{jsonpath this '$.foo'}}}\",\"responseBaz\":{{{jsonpath this '$.baz'}}} ,\"responseBaz2\":\"Bla bla {{{jsonpath this '$.foo'}}} bla bla\"}",
 ```

File: spring-cloud-contract-spec/src/main/groovy/org/springframework/cloud/contract/spec/ContractTemplate.groovy [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	 * @param key
+	 */
+	String header(String key)
+
+	/**
+	 * Returns the template for retrieving the nth value of a request header e.g. {{{ request.headers.X-Request-Id.[5] }}}
+	 * @param key
+	 * @param index
+	 */
+	String header(String key, int index)
```
Comment (cah-andrewfitzgerald) [inline spring-cloud-contract-spec/src/main/groovy/org/springframework/cloud/contract/spec/ContractTemplate.groovy]: you noted that the query param is zero indexed, might want to do that here too

File: spring-cloud-contract-spec/src/main/groovy/org/springframework/cloud/contract/spec/internal/FromRequest.groovy [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	 */
+	DslProperty header(String key) {
+		return new DslProperty(template.header(key))
+	}
+
+	/**
+	 * nth value of a request header e.g. request.headers.X-Request-Id
+	 * @param key
+	 */
+	DslProperty header(String key, int index) {
```
Comment (cah-andrewfitzgerald) [inline spring-cloud-contract-spec/src/main/groovy/org/springframework/cloud/contract/spec/internal/FromRequest.groovy]: same thing about documenting that it's zero-indexed

File: spring-cloud-contract-spec/src/main/groovy/org/springframework/cloud/contract/spec/internal/HandlebarsContractTemplate.groovy [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	}
+
+	@Override
+	String url() {
+		return wrapped("request.url")
+	}
+
+	@Override
+	String query(String key) {
+		return wrapped("request.query.${key}.[0]")
```
Comment (cah-andrewfitzgerald) [inline spring-cloud-contract-spec/src/main/groovy/org/springframework/cloud/contract/spec/internal/HandlebarsContractTemplate.groovy]: could call the overloaded version here to reduce some magic string duplication, but that's just a style thing

**Comment by marcingrzejszczak** [comment]
> The only thing i can think of is that it would be nice if the response body had nicer formatting in the Wiremock stub, but that would probably be more realistic if it wasn't an escaped string.

No way to change this. We're using WireMock's API and this is how it stores data

**Pull request merged**